examples/step-19/step-19.cc

```diff
@@ -0,0 +1,728 @@
+                 ++particle, ++particle_index)
+              {
+                const Tensor<1, dim> E = field_gradients[particle_index];
+
+                ArrayView<double> properties = particle->get_properties();
+                Tensor<1, dim>    old_velocity;
+                for (unsigned int j = 0; j < dim; ++j)
+                  old_velocity[j] = properties[j];
+
+                const Tensor<1, dim> acceleration =
+                  Constants::electron_charge / Constants::electron_mass * E;
+
+                const Tensor<1, dim> new_velocity =
+                  old_velocity + acceleration * dt;
+
+                particle->set_properties(make_array_view(new_velocity));
+
+                // set new location
+                // TODO: This should be more accurate if we used x_k +
+                // dt*(v_k+v_{k+1})
+                //       but that doesn't appear to work. investigate why
+                particle->set_location(particle->get_location() +
+                                       dt * old_velocity);
+              }
+          }
+        }
+
+    // Find the cells that the particles moved to
+    particle_handler.sort_particles_into_subdomains_and_cells();
+  }
+
+
+  // @sect4{CathodeRaySimulator::update_timestep_size}
+
+  template <int dim>
+  void CathodeRaySimulator<dim>::update_timestep_size()
+  {
+    // We need to respect a CFL condition whereby particles can not move further
+    // than one cell. Compute their speed and divide the cell size
+    // by that speed for all particles, then take the minimum and use that
+    // as next time step.
+    if (time.get_step_number() > 0)
+      {
+        double min_cell_size_over_velocity = std::numeric_limits<double>::max();
+
+        for (const auto &cell : dof_handler.active_cell_iterators())
+          if (particle_handler.n_particles_in_cell(cell) > 0)
+            {
+              const double cell_size = cell->minimum_vertex_distance();
+
+              double max_particle_velocity(0.0);
+
+              for (const auto &particle :
+                   particle_handler.particles_in_cell(cell))
+                {
+                  Tensor<1, dim> velocity;
+                  for (unsigned int i = 0; i < dim; ++i)
+                    velocity[i] = particle.get_properties()[i];
+
+                  max_particle_velocity =
+                    std::max(max_particle_velocity, velocity.norm());
+                }
+
+              if (max_particle_velocity > 0)
+                min_cell_size_over_velocity =
+                  std::min(min_cell_size_over_velocity,
+                           cell_size / max_particle_velocity);
+            }
+
+        constexpr double c_saftey = 0.5;
+        time.set_desired_next_step_size(c_saftey * 0.5 *
+                                        min_cell_size_over_velocity);
+      }
+    else
+      {
+        QTrapez<dim>  vertex_quadrature;
+        FEValues<dim> fe_values(fe, vertex_quadrature, update_gradients);
```

Comment:
Does it matter to add the mapping here? Just to prevent possible accidents if this is copied into other codes.

Reply:
I don't have a mapping in this program. There are other places where we would have to do the same -- e.g., interpolate_boundary_values. Let's keep it simple here.